### DISCUSSION TITLE
is there any list or link describe how to conver convert-ssl-ciphers to curl format

### DISCUSSION THREAD

**Discussion by Shawn71** [curl/curl#8760] [Q&A] [answered]
is there any list or link describing how to convert SSL/TLS ciphers to curl format? 
for instance :

I plan to use cipher suite: aes256 and sha384.but I don't know what is it in curl format, is there any list to describe this? thanks 

[root@centos7 ~]# curl https://www.howsmyssl.com/a/check --ciphers AES_256_GCM_SHA384
curl: (59) Unknown cipher in list: AES_256_GCM_SHA384

**Comment by bagder** [comment, marked as answer]
The names are [listed here](https://curl.se/docs/ssl-ciphers.html) and as you can see there, they depend on what TLS backend your curl uses.

**Reply by Shawn71** [reply]
thank you so much !!!! it works. I have two more questions:
question1:
if I use centos8 and don't install any curl or TLS backend, just use the curl by default, how can I ensure what TLS backend use?

question2: if I specify curl using ECDHE-RSA-AES128-SHA256 
root@ubuntu1:/usr/local/src/curl/src# ./curl https://www.baidu.com --ciphers ECDHE-RSA-AES128-SHA256
curl: (35) error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure

but when I do traffic capture ,I can see below 5 cipher suite:
![image](https://user-images.githubusercontent.com/19188604/165686267-21499023-67e1-4993-9638-1dac46f10dd9.png)


how did the rest of the four cipher suites come here?

**Reply by bagder** [reply]
If curl can speak TLS, it means it is built with a TLS backend. The specific one was selected by the person who built your curl binary: someone on centos. Knowing CentOS, you probably have a curl version that is close to a decade old and you probably use something outdated, like the NSS backend. `curl -V` sort of shows what TLS backend it uses, but only in a somewhat hidden way as it shows TLS libraries and their versions on the first line so you need to be able to spot that.

> question2

I don't know.

**Reply by Shawn71** [reply]
really helpful for the answer, thank you so much, no more questions for now.

**Comment by Shawn71** [comment]
Hi  

I just run below command and plan to use curl to upload the file to the server.

command:
curl -k -v --trace-time -u admin:admin -X POST http://file.kaixinduole.com//CreateFileUploadAndDownLoad/servlet/UploadHandleServlet  -F"file=@NTUSER.DAT" -b " JSESSIONID=65E7FC8E5875FFA070A237AB03BB2FC4; kodUserLanguage=zh-CN; kodUserID=1; KOD_SESSION_SSO=5rdqsqgbldl5doukdfhvgdegi5; KOD_SESSION_ID_f865d=ccv29ti4p3fk5409fevpuvfm57; X-CSRF-TOKEN=M2vKIIWvzVhjiq7hyeTI; HOST=http%3A//file.kaixinduole.com/; APP_HOST=http%3A//file.kaixinduole.com/; kodVersionCheck=check-at-1651159568" -H "Content-Type: multipart/form-data" -H "host:file.kaixinduole.com:8080"

Note: Unnecessary use of -X or --request, POST is already inferred.
08:39:19.135678 *   Trying 52.131.41.69:80...
08:39:19.137825 * TCP_NODELAY set
08:39:19.173339 * Connected to file.kaixinduole.com (52.131.41.69) port 80 (#0)
08:39:19.175196 * Server auth using Basic with user 'admin'
08:39:19.176794 > POST //CreateFileUploadAndDownLoad/servlet/UploadHandleServlet HTTP/1.1
08:39:19.176794 > Host:file.kaixinduole.com:8080
08:39:19.176794 > Authorization: Basic YWRtaW46YWRtaW4=
08:39:19.176794 > User-Agent: curl/7.68.0
08:39:19.176794 > Accept: */*
08:39:19.176794 > Cookie:  JSESSIONID=65E7FC8E5875FFA070A237AB03BB2FC4; kodUserLanguage=zh-CN; kodUserID=1; KOD_SESSION_SSO=5rdqsqgbldl5doukdfhvgdegi5; KOD_SESSION_ID_f865d=ccv29ti4p3fk5409fevpuvfm57; X-CSRF-TOKEN=M2vKIIWvzVhjiq7hyeTI; HOST=http%3A//file.kaixinduole.com/; APP_HOST=http%3A//file.kaixinduole.com/; kodVersionCheck=check-at-1651159568
08:39:19.176794 > Content-Length: 3145930
08:39:19.176794 > Content-Type: multipart/form-data; boundary=------------------------47c6d5ca88f99534
08:39:19.176794 > Expect: 100-continue
08:39:19.176794 >
08:39:19.208769 * Mark bundle as not supporting multiuse
08:39:19.209526 < HTTP/1.1 100 Continue
08:39:19.210972 * read function returned funny value
08:39:19.211677 * Closing connection 0
curl: (26) read function returned funny value

meanwhile I do the tcpdump ,please see the below

POST //CreateFileUploadAndDownLoad/servlet/UploadHandleServlet HTTP/1.1
Host:file.kaixinduole.com:8080
Authorization: Basic YWRtaW46YWRtaW4=
User-Agent: curl/7.68.0
Accept: */*
Cookie:  JSESSIONID=65E7FC8E5875FFA070A237AB03BB2FC4; kodUserLanguage=zh-CN; kodUserID=1; KOD_SESSION_SSO=5rdqsqgbldl5doukdfhvgdegi5; KOD_SESSION_ID_f865d=ccv29ti4p3fk5409fevpuvfm57; X-CSRF-TOKEN=M2vKIIWvzVhjiq7hyeTI; HOST=http%3A//file.kaixinduole.com/; APP_HOST=http%3A//file.kaixinduole.com/; kodVersionCheck=check-at-1651159568
Content-Length: 3145930
Content-Type: multipart/form-data; boundary=------------------------15e98e1b96f8a4fb
Expect: 100-continue

HTTP/1.1 100 Continue

--------------------------15e98e1b96f8a4fb
Content-Disposition: form-data; name="file"; filename="NTUSER.DAT"
Content-Type: application/octet-stream

HTTP/1.1 400 Bad Request
Date: Thu, 28 Apr 2022 16:00:56 GMT
Server: Apache/2.4.17 (Unix) OpenSSL/1.0.1p PHP/5.6.14 mod_perl/2.0.8-dev Perl/v5.16.3
Vary: accept-language,accept-charset
Accept-Ranges: bytes
Connection: close
Content-Type: text/html; charset=utf-8
Content-Language: en

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Bad request!</title>
<link rev="made" href="mailto:you@example.com" />
<style type="text/css"><![CDATA[/*></style>
</head>

<body>
    Bad request!
<p>


    Your browser (or proxy) sent a request that
    this server could not understand.

the reason why server respond http 400 to the client is the curl client have the http headers: Expect: 100-continue , so that the server respond it with 100 response code.
then curl send below content:
--------------------------15e98e1b96f8a4fb
Content-Disposition: form-data; name="file"; filename="NTUSER.DAT"
Content-Type: application/octet-stream

so that server does not understand what does client say ,so that give the client http 400.

my question is:

is there any methon to make curl not send "Expect: 100-continue" header and combine the two part below together that send to the server ?

//////////////////////////////////////////////this is the first part/////////////////////////////////////////
POST //CreateFileUploadAndDownLoad/servlet/UploadHandleServlet HTTP/1.1
Host:file.kaixinduole.com:8080
Authorization: Basic YWRtaW46YWRtaW4=
User-Agent: curl/7.68.0
Accept: */*
Cookie:  JSESSIONID=65E7FC8E5875FFA070A237AB03BB2FC4; kodUserLanguage=zh-CN; kodUserID=1; KOD_SESSION_SSO=5rdqsqgbldl5doukdfhvgdegi5; KOD_SESSION_ID_f865d=ccv29ti4p3fk5409fevpuvfm57; X-CSRF-TOKEN=M2vKIIWvzVhjiq7hyeTI; HOST=http%3A//file.kaixinduole.com/; APP_HOST=http%3A//file.kaixinduole.com/; kodVersionCheck=check-at-1651159568
Content-Length: 3145930
Content-Type: multipart/form-data; boundary=------------------------15e98e1b96f8a4fb
Expect: 100-continue **# expect not to send this http header**
///////////////////////////////////////////////////////////////////////////////////////////////////////////////



//////////////////////////////////////////////this is the second part/////////////////////////////////////////
--------------------------15e98e1b96f8a4fb
Content-Disposition: form-data; name="file"; filename="NTUSER.DAT"
Content-Type: application/octet-stream


///////////////////////////////////////////////////////////////////////////////////////////////////////////////

many thanks

**Reply by jay** [reply]
Expect may not be the reason for the bad request reply. You can remove the Expect header by adding `-H Expect:` to the command.  See also https://everything.curl.dev/http/post/expect100
